### PR TITLE
Refactored to derive

### PR DESCRIPTION
**Pull request by mbround18**
# Description

## Contributions

- Upgraded Deps
- Added clap derive (and upgraded to clap 4.x)
- Refactored to use clap derive across the whole code for ease of use

## Checklist

- [ ] I added one or multiple labels which best describes this PR.
- [ ] I have tested the changes locally.
- [ ] This PR has a reviewer on it.
- [ ] I have validated my changes in a docker container and on Ubuntu. (Only needed for Odin or Docker Changes)

### CODE REVIEW

File: src/odin/cli.rs [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+  },
+
+  /// Installs Valheim with steamcmd
+  Install {},
+
+  /// Starts Valheim
+  Start {},
+
+  /// Stops Valheim
+  Stop {},
```
Comment (CosmicHorrorDev) [inline src/odin/cli.rs:59]:
```suggestion
  Stop,
```

File: src/odin/cli.rs [hunk 2]
```diff
@@ -0,0 +1,130 @@
+use clap::{Parser, Subcommand};
+
+#[derive(Parser)]
+#[command(author, version)]
+#[command(propagate_version = true)]
+pub struct Cli {
```
Comment (CosmicHorrorDev) [inline src/odin/cli.rs:6]: I would recommend adding

```rust
#[cfg(test)]
mod tests {
  use super::*;

  use clap::CommandFactory;

  #[test]
  fn asserts() {
    Cli::command().debug_assert();
  }
}
```

to the bottom of the file. `clap` gave up trying to enforce _everything_ at compile time, so they added `.debug_assert()` as a way to automatically test for some issues. This picked up on the conflicting names with some of the short args in `configure`

File: src/odin/Cargo.toml [hunk 3]
```diff
@@ -30,10 +30,10 @@ path = "lib.rs"
 handlebars = "4"
 dotenv = "0.15"
 log = "0.4"
-clap = { version = "2" , features = ["yaml"] }
+clap = { version = "4.1.1", features = [ "derive" ] }
```
Comment (CosmicHorrorDev) [inline src/odin/Cargo.toml:33]:
```suggestion
clap = { version = "4.1.1", features = [ "derive", "env" ] }
```

File: src/odin/cli.rs [hunk 4]
```diff
@@ -0,0 +1,130 @@
...
+  },
+
+  /// Sends a notification to the provided webhook.
+  Notify {
+    /// Title of the message block (required by discord & generic webhook, automatically supplied, default: "Broadcast")
+    #[arg(short, long, value_name = "TITLE", default_value_t = {"Broadcast".to_string()})]
+    title: String,
+
+    /// Message to send to the webhook.
+    #[arg(short, long, value_name = "MESSAGE", default_value_t = {"Test Notification".to_string()})]
```
Comment (CosmicHorrorDev) [inline src/odin/cli.rs:92]:
```suggestion
    #[arg(default_value_t = {"Test Notification".to_string()})]
```

File: src/odin/cli.rs [hunk 5]
```diff
@@ -0,0 +1,130 @@
...
+    /// Sets the webhook to send a notification to, (Can be set with ENV variable WEBHOOK_URL)
+    #[arg(short, long, value_name = "WEBHOOK_URL")]
+    webhook_url: Option<String>,
+    // Sets the webhook to include the server's public IP in notifications, (Can be set with ENV variable WEBHOOK_INCLUDE_PUBLIC_IP)
+    // #[arg(
+    //   short,
+    //   long = "webhook_include_public_ip",
+    //   value_name = "WEBHOOK_INCLUDE_PUBLIC_IP"
+    // )]
+    // webhook_include_public_ip: bool,
```
Comment (CosmicHorrorDev) [inline src/odin/cli.rs:104]: Is this supposed to be commented out?

**Review by CosmicHorrorDev — CHANGES_REQUESTED** [review]
Lots of small stuff. Some important and some less so

Great to see things switched over to `clap` derive though :tada:

File: src/odin/main.rs [hunk 6]
```diff
@@ -19,41 +19,92 @@ pub mod utils;
...
+  //   ("stop", sub_m) => commands::stop::invoke(sub_m.unwrap()),
+  //   ("backup", sub_m) => commands::backup::invoke(sub_m.unwrap()),
+  //   ("notify", sub_m) => commands::notify::invoke(sub_m.unwrap()),
+  //   ("update", sub_m) => commands::update::invoke(sub_m.unwrap()),
+  //   ("mod:install", sub_m) => commands::install_mod::invoke(sub_m.unwrap()),
+  //   ("status", sub_m) => commands::status::invoke(sub_m.unwrap()),
+  //   _ => {
+  //     panic!("No Command Launched!");
+  //   } // Either no subcommand or one not tested for...
+  // }
```
[the remark this answers is not in the document]
Comment (mbround18) [reply src/odin/main.rs:109]: yee

File: src/odin/cli.rs (hunk 5 above)
Comment (mbround18) [reply src/odin/cli.rs:115]: yes

**Review by CosmicHorrorDev — DISMISSED** [review]
Everything looks good enough to me :+1:

**Comment by mbround18** [comment]
Will be testing with `mbround18/valheim:sha-9ef47e4` soon for validation before merging

**Comment by mbround18** [comment]
Damn invalid values for public flag, changing it up to keep the syntax

**Comment by mbround18** [comment]
This PR will be merged Tuesday the 21st

**Pull request merged**